src/test/java/javazoom/jl/player/JavaSoundAudioDeviceTest.java

```diff
@@ -23,6 +25,7 @@ public class JavaSoundAudioDeviceTest {
     @Test
     void test() throws Exception {
         FactoryRegistry factoryRegistry = FactoryRegistry.systemRegistry();
+        factoryRegistry.addFactory(new MyJavaSoundAudioDeviceFactory());
```

Comment:
jlayer uses service provider mechanizm (sorry not documented), so it's no need to do this.
just include `target/test-classes` in classpath

Reply:
Hm. In the module-info.java I need to enumerate all service providers. Since the compiled module-info will be in the release jar, I cannot include the MyJavaSoundAudioDeviceFactory from the test-classes there. But this means the service loader only detects 2 service providers, but expects 3 - therefore I added the third one explicitly.
Without it ```mvn test``` fails.

Reply:
this is the hint???

https://www.javainthebox.com/2018/08/project-jigsaw.html (the article is in japanese, please use translator)

Reply:
No.
I already know how to make a ServiceLoader work for Jigsaw modules - I created the required entries in the module-info.java

The problem is that the service provider added to module-info.java must be accessible to third party modules, which means you cannot include service providers only available in the test sources. 
JLayer has
- 2 service providers in the regular sources 
- 1 service provider in the test sources
The unit tests expects all 3, but will only find those mentioned in Jigsaw module-info.java. That is why I added that line to the unit test, to make it work in modular and non-modular cases.